crates/core/tedge_mapper/src/c8y/converter.rs

```diff
@@ -224,30 +226,75 @@ where
-            let cumulocity_event_json = serde_json::to_string(&c8y_event)?;
-            let json_mqtt_topic = Topic::new_unchecked(C8Y_JSON_MQTT_EVENTS_TOPIC);
+                c8y_event = C8yCreateEvent::try_from(tedge_event)?;
+                // If the message contains extra fields other than `text` and `time`, convert to Cumulocity JSON
+                let cumulocity_event_json = serde_json::to_string(&c8y_event)?;
 
-            Message::new(&json_mqtt_topic, cumulocity_event_json)
+                Message::new(
+                    &Topic::new_unchecked(C8Y_JSON_MQTT_EVENTS_TOPIC),
+                    cumulocity_event_json,
+                )
+            }
+            None => {
+                let tedge_event =
+                    ThinEdgeEvent::try_from(input.topic.name.as_str(), input.payload_str()?, None)?;
+                c8y_event = C8yCreateEvent::try_from(tedge_event)?;
+
+                // If the message doesn't contain any fields other than `text` and `time`, convert to SmartREST
+                if c8y_event.extras.is_empty() {
+                    let smartrest_event = Self::serialize_to_smartrest(&c8y_event)?;
+                    let smartrest_topic = Topic::new_unchecked(SMARTREST_PUBLISH_TOPIC);
+
+                    Message::new(&smartrest_topic, smartrest_event)
+                } else {
+                    // If the message contains extra fields other than `text` and `time`, convert to Cumulocity JSON
+                    let cumulocity_event_json = serde_json::to_string(&c8y_event)?;
+                    let json_mqtt_topic = Topic::new_unchecked(C8Y_JSON_MQTT_EVENTS_TOPIC);
+
+                    Message::new(&json_mqtt_topic, cumulocity_event_json)
+                }
+            }
         };
 
         // If the MQTT message size is well within the Cumulocity MQTT size limit, use MQTT to send the mapped event as well
         if input.payload_bytes().len() < self.size_threshold.0 {
-            Ok(vec![message])
+            vec.push(message);
+            Ok(vec)
+        // If the message size is larger than the MQTT size limit, use HTTP to send the mapped event
         } else {
-            // If the message size is larger than the MQTT size limit, use HTTP to send the mapped event
-            let _ = self.http_proxy.send_event(c8y_event).await?;
-            Ok(vec![])
+            match child_id {
+                Some(id) => {
+                    if self.children.contains(&id) {
+                        let _ = self.http_proxy.send_event(c8y_event).await?;
+                        Ok(vec![])
+                    } else {
+                        Err(ConversionError::ChildDeviceNotRegistered { id: id.to_string() })
+                    }
+                }
+                // Parent device
+                None => {
+                    let _ = self.http_proxy.send_event(c8y_event).await?;
+                    Ok(vec![])
+                }
+            }
```

Comment:
```suggestion
            if let Some(id) = child_id {
                if self.children.contains(&id) {
                    let _ = self.http_proxy.send_event(c8y_event).await?;
                    Ok(vec![])
                } else {
                    Err(ConversionError::ChildDeviceNotRegistered { id: id.to_string() })
                }
            } else {
                // Parent device
                let _ = self.http_proxy.send_event(c8y_event).await?;
                Ok(vec![])
            }
```

This is just a suggestion, your code looks absolutely fine! But as I've seen, you're using `match` a lot even if you're handling `Option<_>` - maybe you just didn't know that you can also do a `if let Some(thing) = optional_thing {}`!

As said in the introductory message: Feel free to ignore me entirely :wink:

Reply:
Thanks for your comments on this PR. This will really help me to learn more (idiomatic) things about Rust.